### DISCUSSION TITLE
Help configuring EAP-PEAP client

### DISCUSSION THREAD

**Discussion by aitzubi** [strongswan/strongswan#400] [Q&A]
Good afternoon,

I need help setting up a strongswan client configuration. Our company is using Microsoft AlwaysOn VPN (IKEv2). Our Windows clients connect to the VPN with no problem using a single certificate and using EAP-PEAP protocol. This certificate constains the userPrincipalName (email address) on the subject and SAN. From this email address, our NPS (Radius) server authenticates the user and allows him to connect to the vpn.

I am unable to find any documentation regarding how to configure this kind of client on linux using Strongswan. We had success using EAP-MSCHAPv2 with Strongswan but we are unable to make it work by only using this certificate and our internal CA certificate.

Could somebody please help us with this issue?

Kind regards,

**Comment by tobiasbrunner** [comment]
What exactly is the issue?

**Reply by aitzubi** [reply]
Hi @tobiasbrunner !

I am unable to find a valid configuration example to how to configure eap-peap client in strongswan using only certificates.

**Reply by tobiasbrunner** [reply]
What difficulty do you have?

**Reply by aitzubi** [reply]
As I said earlier, I can´t find an example on how to configure this scenario on strongswan. Can you provide me with an example @tobiasbrunner ? Thanks!

**Reply by tobiasbrunner** [reply]
Maybe start with the _eap-tls_ example from [here](https://wiki.strongswan.org/projects/strongswan/wiki/UsableExamples) and change the authentication method to _eap-peap_. I'm not entirely sure if that works of if the plugin expects an inner EAP method (if you only want certificate authentication via EAP, why don't you use EAP-TLS? EAP-PEAP is really only to tunnel a potentially insecure username/password method like EAP-MSCHAPv2 in TLS.)

**Reply by aitzubi** [reply]
Thanks! I will try that. 

I followed this guide for pfsense eap-tls in windows and it works! https://docs.netgate.com/pfsense/en/latest/recipes/ipsec-mobile-ikev2-eap-tls.html

Maybe there is hope after all ;)

**Reply by aitzubi** [reply]
@tobiasbrunner I can get it to work. I am using this configuration:

https://openwrt.org/docs/guide-user/services/vpn/strongswan/roadwarrior#ipsecconf

The problem is that our VPN has different endpoints (servers) and our hostname is basically a load balancer. So I don´t have (or don´t know) how to specify a server certificate. I do have a CA certificate (chain) but it is not working :(

**Reply by tobiasbrunner** [reply]
Can't help you without logs and configs.

**Reply by aitzubi** [reply]
Hi @tobiasbrunner ,

First of all, thanks for your help :)

I will try to explain our scenario. We have 3 firewalls, each one on different regions of our country, that provide IKEv2 VPN. Each one of this FWs has a self-signed certificate pointing to its specific hostname (CN) and our cloud load balancer fqdn (SAN).

We use this setup to load balance connections of our users between regions. This setup works without any issues in Windows 10 PCs where each user gets a personal certificate that contains its email (E) and its userprincipalname attribute (SAN). In windows, we just choose to trust our internal CA when configuring the EAP-TLS connection and it works.

We want to replicate the same setup in linux (Ubuntu) using strongswan. For this, I created and exported a certificate for myself from our internal windows CA. I transferred it to a Ubuntu VM and tried this config:

ipsec.conf

```
config setup
        # strictcrlpolicy=yes
        # uniqueids = no

# Add connections here.

conn %default
        keyexchange=ikev2
        ike=aes256-aes128-sha256-sha1-modp3072-modp2048
        esp=aes128-aes256-sha256-modp3072-modp2048,aes128-aes256-sha256
        left=%any
        leftauth=eap-tls
        leftcert=/pathto/usercert.pem
        leftid=%any
        right=loadbalancer.example.com
        eap_identity="myemail@example.com"
        auto=add

conn rwPUBKEY
        rightauth=pubkey
        rightid="loadbalancer.example.com"
```

ipsec.secret

```
# This file holds shared secrets or RSA private keys for authentication.

# RSA private key for this host, authenticating it to any other host
# which knows the public part.

: RSA /pathto/usercert.key
```

This is the log:

```
initiating IKE_SA rwPUBKEY[1] to 111.111.111.2
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 10.0.2.15[500] to 111.111.111.2[500] (1528 bytes)
received packet: from 111.111.111.2[500] to 10.0.2.15[500] (38 bytes)
parsed IKE_SA_INIT response 0 [ N(INVAL_KE) ]
peer didn't accept DH group MODP_3072, it requested MODP_2048
initiating IKE_SA rwPUBKEY[1] to 111.111.111.2
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 10.0.2.15[500] to 111.111.111.2[500] (1400 bytes)
received packet: from 111.111.111.2[500] to 10.0.2.15[500] (485 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(HTTP_CERT_LOOK) CERTREQ V ]
received unknown vendor ID: dd:47:7b:3d:56:b7:72:0c:b4:21:05:71:f6:d2:06:80
selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
local host is behind NAT, sending keep alives
received 1 cert requests for an unknown ca
no IDi configured, fall back on IP address
establishing CHILD_SA rwPUBKEY{1}
generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
sending packet: from 10.0.2.15[4500] to 111.111.111.2[4500] (400 bytes)
received packet: from 111.111.111.2[4500] to 10.0.2.15[4500] (36 bytes)
payload type NOTIFY was not encrypted
could not decrypt payloads
integrity check failed
IKE_AUTH response with message ID 1 processing failed
```

Being 111.111.111.2 the FW node IP.

After doing some research, in every example that I find about this type of connection I find that you need the servers certificate in order to connect but in our case I cannot specify a server certificate due to having 3 nodes. Is it possible to trust directly the certificate based on our CA certificate? 

BTW, I put our CA and Issuing CAs certificates on /etc/ipsec.d/cacerts folder.

Could you or somebody help me with this setup?

Kind regards,

**Reply by tobiasbrunner** [reply]
This fails in a strange way (unencrypted notify) before certificates even become relevant. So check the server log to see what it didn't like about the IKE_AUTH request.

Also note that the fallback to the IP address for IDi shouldn't happen if `leftcert` was correctly specified (the local identity should default to the certificate's subject DN if `leftid` isn't specified or isn't confirmed by the certificate). So check your local log when the daemon starts and the configs are loaded to see why the certificate isn't loaded properly.
